### ISSUE TITLE
Complex Yukawa Couplings

### ISSUE THREAD

**Issue by dhjjacob** [GM2Calc/GM2Calc#13]
I recently tried to use complex Yukawa couplings for the GTHDM.  I noticed in the python interface I couldn't insert complex numbers into the `basis.Pi_f` Yukawa matrices.  I got the error:

`    basis.Pi_u.__setitem__(2,2,std.complex['double'](1.15,0.3))`
`TypeError: none of the 3 overloaded methods succeeded. Full details:`
`  double& Eigen::DenseBase<Eigen::Matrix<double,3,3,0,3,3> >::operator()(long index) =>`
`    TypeError: takes at most 1 arguments (2 given)`
`  double& Eigen::DenseBase<Eigen::Matrix<double,3,3,0,3,3> >::operator[](long index) =>`
`    TypeError: takes at most 1 arguments (2 given)`
`  double& Eigen::DenseBase<Eigen::Matrix<double,3,3,0,3,3> >::operator()(long row, long col) =>`
`    TypeError: must be real number, not complex<double>`

Checking in the THDM.hpp file, I see in the definition of the `Mass_basis` and `Gauge_basis`:

`   Eigen::Matrix<double,3,3> Delta_u{Eigen::Matrix<double,3,3>::Zero()};`
`   Eigen::Matrix<double,3,3> Delta_d{Eigen::Matrix<double,3,3>::Zero()};`
`   Eigen::Matrix<double,3,3> Delta_l{Eigen::Matrix<double,3,3>::Zero()};`
`   Eigen::Matrix<double,3,3> Pi_u{Eigen::Matrix<double,3,3>::Zero()};`
`   Eigen::Matrix<double,3,3> Pi_d{Eigen::Matrix<double,3,3>::Zero()};`
`   Eigen::Matrix<double,3,3> Pi_l{Eigen::Matrix<double,3,3>::Zero()};`

Whereas internally in the THDM_parameters.hpp file we have:

`   Eigen::Matrix<std::complex<double>,3,3> Pi_u{Eigen::Matrix<std::complex<double>,3,3>::Zero()};`
`   Eigen::Matrix<std::complex<double>,3,3> Pi_d{Eigen::Matrix<std::complex<double>,3,3>::Zero()};`
`   Eigen::Matrix<std::complex<double>,3,3> Pi_l{Eigen::Matrix<std::complex<double>,3,3>::Zero()};`

So it seems to me that setting the Yukawa matrices to `double`s in the THDM.hpp file is preventing GM2Calc2 form using complex yukawa couplings.  To test that the problem was not just in the python interface, I also tried inserting imaginary valuesinto the yukawas in the Mathematica interface.  These imaginary values were ignored, as if only the real parts of the Yukawa couplings were included.  

I tried to complete a quick solution by changing the definition of the `Pi_f` matrices in the THDM.hpp file from `double` to `std::complex<double>`.  However this runs into issues with the C/C++ interface.  

So to summarize, currently it seems that GM2Calc2 cannot use complex values in the Yukawa couplings.  This seems to be because of the variables that we use to take inputs from the user.

**Comment by Expander** [comment]
You are right, currently GM2Calc does not deal with complex Yukawa couplings in the 2HDM. The reason is that for GM2Calc 2.0.0 (which introduces the 2HDM) we were not absolutely sure if the implemented expressions are correct when the couplings are complex. For this reason the official C/C++ interface in `THDM.hpp` defines only real coupling matrices, as you said.

In principle this can be generalized. However, we'd need to make sure that all expressions are correctly taking the imaginary parts of the couplings into account. I'm just not sure about this. Do you have some insights into the expressions regarding the imaginary parts?

**Comment by pathron** [comment]
Alex I only have a vague memory of this.  I think the code internally has some expressions that generalise  to allow complex values while reproducing the know result for real yukawas. But there was some doubts that these are correct for all the two-loop diagrams ?  Do you recall what was / was not sure about the result and how to generalise?

**Comment by Expander** [comment]
Sorry, I don't recall exactly which expressions are already general and which aren't. From the top of my head I just remember that the following quantities may be complex in general:

* internally: `lambda5`, `lambda6`, `lambda7`, `Gammau`, `Gammad`, `Gammae`, `Piu`, `Pid`, `Pie`, `Yu`, `Yd`, `Ye` (and as a consequence the mass and mixing matrices)
* externally `lambda5`, `lambda6`, `lambda7`, `Deltau`, `Deltad`, `Deltal`, `Piu`, `Pid`, `Pil` (and maybe even the `zetau`, `zetad`, `zetal`)

For me this illustrates that the (correct) generalization to complex parameters is not trivial and may involve its own little research project. At least, just allowing for complex input parameters does not necessarily mean that the results are fully correct w.r.t. the contributions from the imaginary parts.

**Comment by dhjjacob** [comment]
For the project @pathron and I are working on, it is sufficient that we simply enable `Pif` to be complex, as we are working in the G2HDM.

**Comment by Expander** [comment]
Do you have a "local" copy of GM2Calc where you want do allow for complex `Pif`? (If that would be possible I'd prefer that.)

**Comment by dhjjacob** [comment]
Hi @Expander, so make the changes just to GM2Calc on my local machine?  Yes that could work.

**Comment by pathron** [comment]
@dhjjacob I think also maintain a fork with this change if its non-trivial work to do.

But...

**Comment by pathron** [comment]
@Expander are you saying the only issue is in translating phases from one basis to another?  I don't think that is a project in itself as its not physical and I can't imagine its that difficult actually.  I was worried that the actual 2-loop calculations done in [1607.06292](https://arxiv.org/abs/1607.06292)  assumed (beyond CKM) Yukawas are real in some way.  I had this vague memory that you and Douglas were asking Dominik to confirm you had the correct generalisation and Dominik said he did not know and that is why we made it so only real ones could be input in the interface. Its possible this vague memory is mixed up with something else or entirely delusional...

**Comment by Expander** [comment]
> @dhjjacob I think also maintain a fork with this change if its non-trivial work to do.

Yes, I think a fork with this single change would technically be best.
